core/store/src/trie/mod.rs

```diff
@@ -615,10 +615,19 @@ impl Trie {
         }
     }
 
-    fn retrieve_node(&self, hash: &CryptoHash) -> Result<TrieNodeWithSize, StorageError> {
+    /// Retrieves decoded node alongside with its raw bytes representation.
+    ///
+    /// Note that because Empty nodes (those which are referenced by
+    /// [`Self::EMPTY_ROOT`] hash) aren’t stored in the database, they don’t
+    /// have a bytes representation.  For those nodes the first return value
+    /// will be `None`.
+    fn retrieve_node(
+        &self,
+        hash: &CryptoHash,
+    ) -> Result<(Option<std::sync::Arc<[u8]>>, TrieNodeWithSize), StorageError> {
         match self.retrieve_raw_node(hash)? {
-            None => Ok(TrieNodeWithSize::empty()),
-            Some((_bytes, node)) => Ok(TrieNodeWithSize::from_raw(node)),
+            None => Ok((None, TrieNodeWithSize::empty())),
+            Some((bytes, node)) => Ok((Some(bytes), TrieNodeWithSize::from_raw(node))),
         }
     }
```

Comment:
I'm unsure of what to do here.

Reply:
It’s just a matter of separating this diff hunk and all the places where retrieve_node is called into a new commit and submitting as separate PR.  It will get this PR smaller.

Reply:
it's not super trivial as far as I see. Need to also modify `descend_into_node`. Can't we just keep it as it is?

Reply:
It should be rather simple mechanical change.  In that separate PR, the change to desdend_into_node will be change from `self.trie.retrieve_node(hash)?` to `self.trie.retrieve_node(hash)?.1`.

Reply:
and what about the `bytes` that are pushed into `nodes`? they aren't available anymore

Reply:
sorry @mina86 but I'm truly not following you on this one

Reply:
There are three separate PRs here:
1. PR changing retrieve_node to return (bytes, TrieNodeWithSize).  All that PR does is changes the return type and updates all the call sites to do `retrieve_node(...)?.1`.  This is quite mechanical and boring change but thanks to that it should be easy to write, review and merge.  This PR is not concerned with the overall feature.
2. PR changing TrieIterator::seek to TrieIterator::seek_prefix and makes changes the iterator such that it stops once everything within the prefix has been traversed.  This can be done by adding a `prefix_boundary:  bool` to `Crumb` and changing Crumb::increment such that if that flag is set, it always moves to Exiting status.  The purpose of this change is to make sure that when iterating over a prefix, the iterator does not fetch spurious nodes.  Currently I believe that returned proof includes unnecessary nodes.  There may be some smarter way of dealing with this but I’m not seeing it just now.
3. Once 1 and 2 are done, this PR can be rebased and will be ready to go.

1 and 2 can be done in parallel.

Reply:
For (2) I've created https://github.com/ComposableFi/nearcore/pull/11 Can you please review it there?

Reply:
For (1) https://github.com/near/nearcore/pull/7535